### ISSUE TITLE
Handle syntax in PEP-612 for user-defined generic classes

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#2449]
**Is your feature request related to a problem? Please describe.**
In PEP-612, there is a section named [User-Defined Generic Classes](https://www.python.org/dev/peps/pep-0612/#user-defined-generic-classes) which details the syntax developers can use to pass arguments to generic classes with `ParamSpec` parameters. Currently, pyright rejects this syntax.

**Describe the solution you'd like**
Here are the relevant code samples from the section in PEP-612:

```python
T = TypeVar("T")
P_2 = ParamSpec("P_2")

class X(Generic[T, P]):
  f: Callable[P, int]
  x: T

def f(x: X[int, P_2]) -> str: ...                    # Accepted
def f(x: X[int, Concatenate[int, P_2]]) -> str: ...  # Accepted
def f(x: X[int, [int, bool]]) -> str: ...            # Accepted
def f(x: X[int, ...]) -> str: ...                    # Accepted
def f(x: X[int, int]) -> str: ...                    # Rejected

class Z(Generic[P]):
  f: Callable[P, int]

def f(x: Z[[int, str, bool]]) -> str: ...   # Accepted
def f(x: Z[int, str, bool]) -> str: ...     # Equivalent
```

I would like to see the following:
* For 3.10 and higher, the above would either be accepted or rejected as noted.
* For versions `>=3.7` and `<3.10`, if `from __future__ import annotations` is used, the above would be accepted or rejected as noted (it should follow the same rules as the union syntax introduced in 3.9).
* For all other versions without postponed evaluation of annotations, the arguments would need to be passed in via strings. I'm still waiting on an answer to [a question](https://github.com/python/typing/discussions/908) I have about the exact syntax that should be accepted in those strings, but I think a good guess would be the following:
```python
def f(x: X[int, '[int, bool]']) -> str: ...            # Accepted
def f(x: Z['[int, str, bool]']) -> str: ...   # Accepted
def f(x: Z['int, str, bool']) -> str: ...     # Possibly accepted, but I could see an argument for rejected as well
```

**Comment by erictraut** [comment]
Thanks for pointing out this omission. I hadn't noticed that section of the PEP previously.

As for quoted expressions, I don't plan to accept any of the above. You would need to enclose the entire annotation in quotes, like this: `"X[int, [int, bool]]"`. This is consistent with the handling of quotes in other type expressions where the quoted section needs to be a valid type expression on its own. For example, you can't use quotes like this: `tuple"[int]"`. So all of the above quoted examples will be rejected.

**Comment by erictraut** [comment]
This functionality will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.180, which I just published. It will also be included in the next pylance release.
